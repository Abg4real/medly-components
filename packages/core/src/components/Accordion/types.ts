import { Props as TitleProps } from './Title';

export interface Props {
    Title?: React.SFC<TitleProps>;
    Content?: React.SFC;
}

export interface AccordionContextProps {
    isActive?: boolean;
    toggle?: () => void;
}
