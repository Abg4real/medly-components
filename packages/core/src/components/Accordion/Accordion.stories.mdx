import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import { CountStyle, ThemeInterface, Data} from './Accordion.stories';
import { Accordion  } from './Accordion';
import { AccordionTheme, defaultTheme } from '@medly-components/theme';



<Meta title="Core/Accordion" component={Accordion} parameters={{ jest: ['Accordion.test.tsx'] }} />

# Accordion

Accordion is a container that can either stand alone or be connected with other components.

#### Basic Accordion
<Preview withToolbar>
    <Story name="BasicAccordion" parameters={{ decorators: [withKnobs] }}>
        <Accordion>
            <Accordion.Title>
                <Text> Uploaded Files</Text>
            </Accordion.Title>
            <Accordion.Content>
                <div>File 1</div>
            </Accordion.Content>
        </Accordion>
    </Story>
</Preview>

#### Accordion with title content
<Preview withToolbar>
    <Story name="AccordionWithCount" parameters={{ decorators: [withKnobs] }}>
        <Accordion>
            <Accordion.Title bgColor={defaultTheme.colors.springTeal[100]}>
                <Data> Uploaded Files</Data>
                <CountStyle textColor={defaultTheme.colors.communityBlue[500]} textWeight="Strong">
                    34
                </CountStyle>
            </Accordion.Title>
            <Accordion.Content>
                <div>File 1</div>
            </Accordion.Content>
        </Accordion>
    </Story>
</Preview>

#### Accordion Theme

<Props of={ThemeInterface} />
