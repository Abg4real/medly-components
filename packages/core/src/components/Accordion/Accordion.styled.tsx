import { styled } from '@medly-components/utils';

export const Section = styled('div')`
    display: flex;
    flex-direction: column;
`;
