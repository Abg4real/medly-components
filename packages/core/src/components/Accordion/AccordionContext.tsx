import { createContext } from 'react';
import { AccordionContextType } from './types';

export const AccordionContext = createContext<AccordionContextType | null>(null);
