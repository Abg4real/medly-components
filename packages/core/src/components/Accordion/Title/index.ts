export { Title as default } from './Title';
export * from './types';

