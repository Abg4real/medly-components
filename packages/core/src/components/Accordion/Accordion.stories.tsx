import React from "react"
import { Accordion } from "./Accordion"
import { AccordionTheme, defaultTheme } from "@medly-components/theme";

export const ThemeInterface: React.SFC<AccordionTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.accordion
};

export const SimpleAccordion = () => {
    return (
        <Accordion title="Uploaded Files" bgColor={defaultTheme.colors.grey[200]} titleColor={defaultTheme.colors.black}> 
            <div>File 1</div>
        </Accordion>

    )
}