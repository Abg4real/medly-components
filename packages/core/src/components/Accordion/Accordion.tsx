import React, { ReactElement, SFC, useCallback, useState } from 'react';
import { Section } from './Accordion.styled';
import { AccordionContext } from './AccordionContext';
import Content from './Content';
import Title from './Title';
import { Props } from './types';

export const Accordion: SFC & Props = React.memo(({ children }) => {
    const [title, content] = React.Children.toArray(children) as ReactElement[],
        [isActive, setIsActive] = useState(false);

    const handleToggle = useCallback(() => setIsActive(!isActive), [isActive]),
        contextValue = React.useMemo(() => ({ isActive, toggle: handleToggle }), [isActive, setIsActive]);

    return (
        <Section>
            <AccordionContext.Provider value={contextValue}>{title}</AccordionContext.Provider>
            {isActive && content}
        </Section>
    );
});

Accordion.displayName = 'Accordion';
Accordion.Title = Title;
Accordion.Content = Content;
