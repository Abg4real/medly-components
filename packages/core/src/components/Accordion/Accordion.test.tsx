import { fireEvent, render, waitFor } from '@test-utils';
import React from 'react';
import { Accordion } from './Accordion';

describe('Accordion component', () => {
    const renderAccordion = () =>
        render(
            <Accordion>
                <Accordion.Title>List Of Components</Accordion.Title>
                <Accordion.Content>Avatar</Accordion.Content>
            </Accordion>
        );
    it('should render properly', () => {
        const { container } = renderAccordion();
        expect(container).toMatchSnapshot();
    });

    it('should render children when accordion is clicked', async () => {
        const { getByText, queryByText } = renderAccordion();
        fireEvent.click(getByText(/List Of Components/));
        await waitFor(() => expect(getByText(/Avatar/)).toBeInTheDocument());
        fireEvent.click(getByText(/List Of Components/));
        expect(queryByText(/Avatar/)).not.toBeInTheDocument();
    });

    it('should invoke handleToggle when accordion is clicked', () => {
        const { getByText, queryByText, container } = renderAccordion();
        fireEvent.click(getByText(/List Of Components/));
        expect(queryByText(/Avatar/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
