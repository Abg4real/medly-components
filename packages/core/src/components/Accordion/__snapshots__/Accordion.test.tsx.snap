// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component should invoke handleToggle when accordion is clicked 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 0.8rem;
}

.c4 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  border-top: solid 0.2rem #eff2f4;
  padding: 1.8rem;
  background-color: transparent;
}

.c3 {
  font-size: 2.4rem;
  cursor: inherit;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
}

.c3 * {
  fill-opacity: 1;
  fill: #607890;
}

.c5 .c2 {
  margin-left: auto;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c1 {
  cursor: pointer;
  padding: 1.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  border: none;
  background-color: #ffffff;
  border-radius: none;
  border-top-left-radius: 0.8rem;
  border-top-right-radius: 0.8rem;
}

.c1 .c2 {
  margin-left: auto;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      List Of Components
      <svg
        class="c2 c3"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.875 9l-3.88 3.88L8.115 9a.996.996 0 1 0-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
    </button>
    <div
      class="c4"
    >
      Avatar
    </div>
  </div>
</div>
`;

exports[`Accordion component should render properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 0.8rem;
}

.c3 {
  font-size: 2.4rem;
  cursor: inherit;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
}

.c3 * {
  fill-opacity: 1;
  fill: #607890;
}

.c1 {
  cursor: pointer;
  padding: 1.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  border: none;
  background-color: #ffffff;
  border-radius: 0.8rem;
  border-top-left-radius: 0.8rem;
  border-top-right-radius: 0.8rem;
}

.c1 .c2 {
  margin-left: auto;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      List Of Components
      <svg
        class="c2 c3"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.875 9l-3.88 3.88L8.115 9a.996.996 0 1 0-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
    </button>
  </div>
</div>
`;
